### PR TITLE
treeHelper macro initial commit

### PR DESCRIPTION
**Pull request by iinjdpa**
Thank you for creating a pull request to contribute to FreeCAD-macros!
To integrate your macro please make sure the following steps are complete:

- [ ] Please check this box if you're not submitting a new macro.   
- [x] Are you submitting a new macro ?  
- [x] Have you followed the ['How to submit a macro'](../README.md#how-to-submit-a-macro) section of the README.md ?  
- [x] Your macro has a [Description](../README.md#macro-description) in its header.  
- [x] Your macro has a [CamelCase name](../README.md#camelcase-macro-name).  
- [x] Your macro is named [appropriately](../README.md#macro-name-specifics).  
- [x] Your macro contains a [Metadata section](../README.md#macro-metadata) that immediately follows the header description.  
- [x] Your macro is Python3/Qt5 compliant and tested on the latest FreeCAD stable and development releases.  
- [x] You're including documentation on how your macro works (bonus: screenshots and/or video on the Wiki)  
- [x] Commit message is [well-written](https://chris.beams.io/posts/git-commit/)  
- [x] Commit message is titled in the following way `[MacroName] Short description`.
- [ ] Optional, write or update the changelog in the macro, from latest to oldest.

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.19 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=34586).

---

### CODE REVIEW

**Comment by luzpaz** [comment]
No spaces in the `[x]` will activate the markdown checklist like so:  
- [x]

**Comment by iinjdpa** [comment]
I edited the markdown. Thank you.

**Comment by luzpaz** [comment]
@iinjdpa do you mind opening a forum thread introducing the macro, perhaps with some screenshots ?

**Comment by iinjdpa** [comment]
@luzpaz I have an old post opened in forum about treeHelper. I will update this post with last functionalities of treeHelper.
I have a youtube video where I have showed treeHelper. I have made an update from youtube video version where treeHelper inherit from QDockWidget.

forum thread url: https://forum.freecadweb.org/viewtopic.php?style=1&f=22&t=63807&p=628636#p628636

File: Utility/treeHelper.FCMacro [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+
+
+class Objdoc(object):
+    def __init__(self,doc):
+        self.ViewObject = mViewObject()
+        self.Name = doc.Name
+        self.Label = doc.Name
+
+class mViewObject(object):
+    def __init__(self):
```
Comment (galou) [inline Utility/treeHelper.FCMacro:40]: Skip `__init__` entirely if not used.

File: Utility/treeHelper.FCMacro [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+__Help__ = "Select a folder, part or body where helper tree must grow. Drag object from main tree to helper tree"
+__Status__ = "Alpha"
+__Requires__ = "Freecad >= v0.19"
+__Communication__ = "https://github.com/iinjdpa/TreeHelper/issues"
+__Files__ = "/treeHelper/ui_treewindow.py,/treeHelper/treeHelper.svg"
+
+
+import FreeCAD as App
+import FreeCADGui as Gui
+from PySide import QtCore, QtGui
```
Comment (galou) [inline Utility/treeHelper.FCMacro:21]: Please state that this is FreeCAD's custom PySide, unfortunately.

File: Utility/treeHelper.FCMacro [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+        self.Name = doc.Name
+        self.Label = doc.Name
+
+class mViewObject(object):
+    def __init__(self):
+        pass
+    def claimChildren(self):
+        # Nombre del documento
+        doc = App.activeDocument()
+        if doc == None:
```
Comment (galou) [inline Utility/treeHelper.FCMacro:45]: `if doc is None:`

File: Utility/treeHelper.FCMacro [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+            doc = FreeCAD.newDocument()
```
Comment (galou) [inline Utility/treeHelper.FCMacro:46]: Why do you create a new document? I though that the macro helps to sort the tree. A new document has no objects to sort.

File: Utility/treeHelper.FCMacro [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+        docName = doc.Name
```
Comment (galou) [inline Utility/treeHelper.FCMacro:47]: Is the variable really necessary? `doc.Name` appears just as short and clear as `docName`.

File: Utility/treeHelper.FCMacro [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+__Wiki__ = ""
+__Icon__ = "/treeHelper/treeHelper.svg"
+__Help__ = "Select a folder, part or body where helper tree must grow. Drag object from main tree to helper tree"
+__Status__ = "Alpha"
+__Requires__ = "Freecad >= v0.19"
+__Communication__ = "https://github.com/iinjdpa/TreeHelper/issues"
+__Files__ = "/treeHelper/ui_treewindow.py,/treeHelper/treeHelper.svg"
+
+
+import FreeCAD as App
```
Comment (galou) [inline Utility/treeHelper.FCMacro:19]: Please use `app` and `gui` for consistency with other macros.

File: Utility/treeHelper.FCMacro [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+            return root_objects
+
+    def update(self):
+        pass
+
+class TreeDrops(QtGui.QTreeWidget):
+    itemDropped = QtCore.Signal()
+    customMimeType = "application/x-qabstractitemmodeldatalist"
+    def __init__(self, parent=None):
+        super(TreeDrops, self).__init__(parent=None)
```
Comment (galou) [inline Utility/treeHelper.FCMacro:85]: Prefer `super().__init__(parent=None)`. I guess that `super().__init__(parent)` is the correct call anyway.

File: Utility/treeHelper.FCMacro [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+            selected_in_FreeCAD = event.source().selectedItems()
+            if selected_in_TreeHelper:
+                selected = selected_in_TreeHelper
+            else:
+                selected = selected_in_FreeCAD
+            # for origin_label in event.source().selectedItems():
+            for origin_label in selected:
+                origin_label = origin_label.text(0)
+                obj_by_label = App.ActiveDocument.getObjectsByLabel(origin_label)
+                if len(obj_by_label) > 1:
```
Comment (galou) [inline Utility/treeHelper.FCMacro:132]: What about `if obj_by_label`?

File: Utility/treeHelper/ui_treewindow.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+# Form implementation generated from reading ui file 'tree_window_.ui'
+#
+# Created by: PyQt5 UI code generator 5.15.7
+#
+# WARNING: Any manual changes made to this file will be lost when pyuic5 is
+# run again.  Do not edit this file unless you know what you are doing.
+
+
+from PySide2 import QtCore, QtGui, QtWidgets
```
Comment (galou) [inline Utility/treeHelper/ui_treewindow.py:11]: Please upload the ui file and replace with `gui.PySideUic.loadUiType('relative_path_to_ui_file')`. Cf. https://github.com/FreeCAD/FreeCAD-macros/wiki/Dynamically-load-graphical-interface-definition.

**Review by galou — CHANGES_REQUESTED** [review]
Thank you for your contribution, a few corrections are required before I can merge:
- Name the macro `TreeHelp.FCMacro` and the icon `TreeHelper.svg`.
- Put the icon at the same level as the macro.
- Remove the trailing slash ('/') in `__Files__`.
- Rename the directory `tree_helper`.
- Use the ui file instead of the generated Python file.
- Other inline comments.

**Comment by galou** [comment]
@iinjdpa do you have the time to have a look at my suggestions?

**Pull request merged**